File: Product_array_except_self.py
# Time COmplexity: O(n)
# Space Complexity: O(1)
# Is this problem on leetcode: Yes
# Approach to solve this problem:

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        result = [1] * len(nums)                       # result is initailized list
        
        prefix = 1
        
        
        for i in range(len(nums)):                    # We are taking left product and right product
            result[i] = prefix
            prefix = prefix * nums[i]
        
        postfix = 1
        for i in range(len(nums)-1,-1,-1):
            result[i] = result[i] * postfix
            postfix = postfix * nums[i]
        return result
        

File: ProductArrayExceptSelf.cpp
// Time Complexity : O(n)

// Space Complexity: O(1)

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> result(nums.size(), 0);
        int ps = 1;
        int ss = 1;  

        result[0] = 1;

        for(int i = 1; i < nums.size(); i++) {
            result[i] = ps * nums[i-1];
            ps = result[i];

        }

        for(int i = nums.size()-2; i >= 0; i--) {
            result[i] = result[i] * ss * nums[i+1];
            ss = ss * nums[i+1];
        }

        return result; 
    }
};